function.dd

```diff
@@ -121,14 +121,61 @@ int foo();
 
 $(H4 $(LNAME2 pure-functions, Pure Functions))
 
-        $(P Pure functions are functions which cannot access global or static, mutable
-        state save through their arguments. This can enable optimizations based on the fact
-        that a pure function is guaranteed to mutate nothing which isn't passed to it,
+    $(P Pure functions are functions which cannot directly access global or static
+        mutable state. It guarantees that a pure function call won't modify any
+        invisible places from the caller side.
+    )
+
+    $(P Different from some other functional programming languages, D's pure
+        function allows modification of caller side state through its mutable
+        parameters.
+
+        ---
+        pure int foo(int[] arr) { arr[] += 1; return arr.length; }
+        int[] a = [1, 2, 3];
+        foo(a);
+        assert(a == [2, 3, 4]);
+        ---
+
+        The minimum guarantee of "pure" is called "weak purity". It's designed to
+        increase modularity of functions.
+    )
+
+    $(P To control mutations, D has the $(D immutable) type qualifier. If all of
+        a $(D pure) function's parameters are $(D immutable) or copied values
+        without any indirections, it can guarantee that the pure function has
+        no side effects.
+    )
+
+        ---
+        struct S { double x; }
+        pure int foo(immutable(int)[] arr, int num, S val)
+        {
+            //arr[num] = 1; // compile error
+            num = 2;        // has no side effect to the caller side
+            val.x = 3.14;   // ditto
+            return arr.length;
+        }
+        ---
+
+    $(P The maximum guarantee of "pure" is called "strong purity". It can enable
+        optimizations based on the fact
+        that a function is guaranteed to not mutate anyhing which isn't passed to it,
         and in cases where the compiler can guarantee that a pure function cannot
         alter its arguments, it can enable full, functional purity (i.e. the guarantee
```

Comment:
Nothing. The current state is my best I can do.
But I can understand that you don't like adding new terms in the language spec.
So, if you write more better text and open a PR, I'll accept it.